presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1876,22 +1965,56 @@ private Scope analyzeWith(Query node, Scope scope)
             return withScope;
         }
 
-        private void analyzeOrderBy(Query node, Scope scope)
+        private void analyzeOrderBy(Query node, Scope orderByScope)
+        {
+            checkState(node.getOrderBy().isPresent(), "orderBy is absent");
+
+            List<SortItem> sortItems = node.getOrderBy().get().getSortItems();
+            analyzeOrderBy(node, sortItems, orderByScope);
+        }
+
+        private List<Expression> analyzeOrderBy(QuerySpecification node, Scope orderByScope, List<Expression> outputExpressions)
         {
-            List<SortItem> items = node.getOrderBy()
-                    .map(OrderBy::getSortItems)
-                    .orElse(emptyList());
+            checkState(node.getOrderBy().isPresent(), "orderBy is absent");
+
+            List<SortItem> sortItems = node.getOrderBy().get().getSortItems();
+
+            if (node.getSelect().isDistinct()) {
+                verifySelectDistinct(node, outputExpressions);
+            }
+
+            return analyzeOrderBy(node, sortItems, orderByScope);
+        }
+
+        private void verifySelectDistinct(QuerySpecification node, List<Expression> outputExpressions)
+        {
+            for (SortItem item : node.getOrderBy().get().getSortItems()) {
+                Expression expression = item.getSortKey();
+
+                if (expression instanceof LongLiteral) {
+                    continue;
+                }
+
+                Expression rewrittenOrderByExpression = ExpressionTreeRewriter.rewriteWith(new OrderByExpressionRewriter(extractNamedOutputExpressions(node.getSelect())), expression);
+                int index = outputExpressions.indexOf(rewrittenOrderByExpression);
+                if (index == -1) {
```

Comment:
Per the discussion in the other issue, we should treat the two invocations as separate calls.